### PR TITLE
Sort series in the same order as legend in graph panel

### LINKED ISSUE #9538: [Feature request] allow sorting stacked series in the same order as legend
I have a stacked chart that uses multi-value template variable to select which series to show:
https://devstats.k8s.io/dashboard/db/companies-stats?orgId=1
I can sort legend by min, max, avg etc.
But stacked series are displayed in alphabetical order, so, for example, Google which is the biggest part of all stacked series is not 1st, Red Hat is not 2nd etc. 

I would like to have an option to sort stacked series in the same order as current order of legend
 
Screenshot:
screen shot 2017-10-16 at 08 42 45

Would be really great to have some checkbox option saying "sort stacked series the same way as legend" or something similar.

### PR DESCRIPTION
**Pull request by bergquist**
closes #9538

### CODE REVIEW

File: public/app/plugins/panel/graph/legend.js [hunk 1]
```diff
@@ -80,13 +80,13 @@ function (angular, _, $) {
...
             panel.legend.sortDesc = null;
-            render();
+            ctrl.events.emit('render');
             return;
           }
 
           panel.legend.sortDesc = !panel.legend.sortDesc;
           panel.legend.sort = stat;
-          render();
+          ctrl.events.emit('render');
```
Comment (torkelo) [inline public/app/plugins/panel/graph/legend.js]: this should be ctrl.render()   (it does the emit)

**Comment by bergquist** [comment]
ready for review!

File: public/app/plugins/panel/graph/graph.ts [hunk 2]
```diff
@@ -343,7 +343,8 @@ function graphDirective($rootScope, timeSrv, popoverSrv, contextSrv) {
         eventManager.addFlotEvents(annotations, options);
         configureAxisOptions(data, options);
 
-        sortedSeries = _.sortBy(data, function(series) { return series.zindex; });
+        sortedSeries = _.sortBy(data, (series) => { return series.zindex; });
+        sortSeries(sortedSeries, data, ctrl.panel);
```
Comment (torkelo) [inline public/app/plugins/panel/graph/graph.ts]: Looks like it sort twice here (when legend sort is enabled), maybe move zindex sort into sortSeries?

File: public/app/plugins/panel/graph/tab_display.html [hunk 3]
```diff
@@ -71,12 +71,14 @@ <h5 class="section-heading">Hover tooltip</h5>
 
 		<div class="section gf-form-group">
 			<h5 class="section-heading">Stacking & Null value</h5>
-			<gf-form-switch class="gf-form" label="Stack" label-class="width-7" checked="ctrl.panel.stack" on-change="ctrl.render()">
+			<gf-form-switch class="gf-form" label="Sort as Legend" label-class="width-8" checked="ctrl.panel.sortAsLegend" on-change="ctrl.render()">
```
Comment (torkelo) [inline public/app/plugins/panel/graph/tab_display.html]: Think a dropdown open named "Stack Order" with two options "By Query" , "By Legend Order". But not sure if that is clearer. Also worth discussing, do we need an option for this? When stacking is this option 95% of the time always desired?

File: public/app/plugins/panel/graph/graph.ts (hunk 2 above)
Comment (bergquist) [reply public/app/plugins/panel/graph/graph.ts]: Should series override (zindex) or sort as legend rank highest?

I think we should execute zindex sorting at the end.

File: public/app/plugins/panel/graph/tab_display.html (hunk 3 above)
Comment (bergquist) [reply public/app/plugins/panel/graph/tab_display.html]: Good point!

I bet there might be cases where this is not the case but then people could use series override to get their series to behave exactly as they want.

File: public/app/plugins/panel/graph/graph.ts (hunk 2 above)
Comment (torkelo) [reply public/app/plugins/panel/graph/graph.ts]: Yes. That might be best. Do it inside sortSeries after any legend sort.

**Comment by torkelo** [comment]
This is looking solid!

**Pull request merged**